docs/source/building/variants.rst

```diff
@@ -0,0 +1,1201 @@
+we need to build binary packages (and any package containing binaries) with
+potentially several variants to support different usage environments. For
+example, using Numpy's C API means that a package must be used with the same
+version of Numpy at runtime that was used at build time.
+
+There has been limited support for this for a long time: Python in both build
+and run requirements resulted in a package that had python pinned to the version
+of Python used at build time (with a corresponding addition to the filename like
+py27). Similar support existed for numpy with the addition of an ``x.x`` pin in
+the recipe after `Conda-build PR
+573 <https://github.com/conda/conda-build/pull/573>`_ was merged. However, there
+has not been general support until conda-build version 3.0, though there have
+been many proposals (`Conda-build issue
+1142 <https://github.com/conda/conda-build/issues/1142>`_).
+
+As of conda-build 3.0, a new configuration scheme has been added, dubbed
+"variants." Conceptually, this decouples pinning values from recipes, replacing
+them with Jinja2 template variables. It adds support for the notion of
+"compatible" pinnings, though that concept is currently still under heavy
+development, to be integrated with ABI compatibility databases, such as `ABI
+Laboratory <https://abi-laboratory.pro/>`_.
+
+Variant input is ultimately a dictionary. They're mostly very flat. Keys are
+made directly available in Jinja2 templates. As a result, keys in the dictionary
+(and in files read into dictionaries) must be valid jinja2 variable names (no ``-``
+characters allowed). This example builds python 2.7 and 3.5 packages in one
+build command:
+
+conda_build_config.yaml like:
+
+.. code-block:: yaml
+
+   python:
+       - 2.7
+       - 3.5
+
+
+meta.yaml contents like:
+
+.. code-block:: yaml
+
+   package:
+       name: compiled-code
+       version: 1.0
+
+   requirements:
+       build:
+           - python {{ python }}
+       run:
+           - python
+
+
+General pinning examples
+------------------------
+
+There are a few characteristic use cases for pinning.  Please consider this a map for the content below.
+
+1. Shared library providing a binary interface. All uses of this library use the
+   binary interface.  It is convenient to apply the same pin to all of your builds.
+   Example: boost
+
+
+conda_build_config.yaml in your HOME folder:
+
+.. code-block:: yaml
+
+  boost:
+    - 1.61
+    - 1.63
+  pin_run_as_build:
+    boost: x.x
+
+
+meta.yaml:
+
+.. code-block:: yaml
+
+   package:
+       name: compiled-code
+       version: 1.0
+
+   requirements:
+       build:
+           - boost  {{ boost }}
+       run:
+           - boost
+
+This example demonstrates several features:
+  * site-wide configuration with a specifically named config file
+    (conda_build_config.yaml in your home folder). More options below in
+    `Creating conda-build variant config files`_.
```

Comment:
Also, I see how the `$HOME/conda_build_config.yaml` is useful but it also opens up a vector for reproducability problems with recipes.  In this example you've given, suppose a user who does not put 
```
pin_run_as_build:
    boost: x.x
```
In their `$HOME/conda_build_config.yaml` builds this recipe.  The behavior of the built package will be dramatically different because the runtime version of boost won't be pinned.

Maybe this is a 'best-practice' issue and not something that conda-build itself needs to address but it seems like there are 'safe' things that you could put in `$HOME/conda_build_config.yaml` and other conda_build_config.yaml values that really should be encapsulated within the recipe.